### DISCUSSION TITLE
Does sqlalchemy.sql.functions really need now() and current_timestamp()?

### DISCUSSION THREAD

**Discussion by davetapley** [sqlalchemy/sqlalchemy#7253] [Ideas]
I just got burned not realizing that `alembic revision --autogenerate` doesn't generate dialect agnostic migrations.

Specifically I used [`sqlalchemy.sql.functions.now()`](https://docs.sqlalchemy.org/en/14/core/functions.html#sqlalchemy.sql.functions.now);
for Postgres this is `now()`, but for MS SQL it is `CURRENT_TIMESTAMP` due to 👇🏻 
https://github.com/sqlalchemy/sqlalchemy/blob/0a57293b844e2164d8a71f6dcfe182bfa0732460/lib/sqlalchemy/dialects/mssql/base.py#L1789-L1790

Since MS SQL doesn't have a `now()` it means that if you `alembic revision --autogenerate` while pointing at Postgres you get a migration file which only works on Postgres, but if you are pointing at MS SQL you get a migration file which works on both.

---

What's frustrating is that:
1. There's already a [`sqlalchemy.sql.functions.current_timestamp()`](https://docs.sqlalchemy.org/en/14/core/functions.html#sqlalchemy.sql.functions.current_timestamp), which generates a `CURRENT_TIMESTAMP` for both dialects (i.e. it generates a dialect agnostic migration file), and:

2. Per [this discussion](https://dba.stackexchange.com/questions/63548/difference-between-now-and-current-timestamp):
      
   > CURRENT_TIMESTAMP, transaction_timestamp() and now() do exactly the same.

---

So my questions are:
Do we really need [`sqlalchemy.sql.functions.now()`](https://docs.sqlalchemy.org/en/14/core/functions.html#sqlalchemy.sql.functions.now)?
What purpose does it serve beyond potentially confusing people? 🙈 

Moreover, even if someone _really_ did want to generate `now()`, they could use [`func`](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.func) anyway:
> Any name can be given to `func`. If the function name is unknown to SQLAlchemy, it will be rendered exactly as is.

**Comment by CaselIT** [comment]
Hi,

> Since MS SQL doesn't have a now() it means that if you alembic revision --autogenerate while pointing at Postgres you get a migration file which only works on Postgres, but if you are pointing at MS SQL you get a migration file which works on both.

what do you mean by this? If the alembic migration uses `func.now()` it will render correctly when executing in mssql. Could you provide a migration example?

**Reply by davetapley** [reply]
>  it will render correctly when executing in mssql

Yes ⬆️  is true if you `--autogenerate` with `sqlalchemy.url = mssql+pymssql...`, i.e. you get:
```
    sa.Column('created', sa.DateTime(timezone=True), server_default=sa.text('CURRENT_TIMESTAMP'), nullable=False),
```

---

But in our development environments we use `sqlalchemy.url = postgresql://`, and then `alembic revision --autogenerate` and check generated migration in. Which (because Postgres) causes:

```
    sa.Column('created', sa.DateTime(timezone=True), server_default=sa.text('now()'), nullable=False),
```

⬆️ this doesn't work in MS SQL.

---

We're now using `functions.current_timestamp()` (which generates `CURRENT_TIMESTAMP` regardless of dialect), but that raises the question of what purpose `functions.now()` serves (other than confusing people like me!).

**Reply by zzzeek** [reply]
Alembic autogenerate has limited options for rendering server_default.  we dont have a full blown "render an entire SQLAlchemy expression in Python code" feature so those necessarily have to come out as text() with the SQL rendered inside them as far as what autogenerate can do automatically.  

For this particular issue, you basically just need to check the rendered migrations as you would normally do anyway (don't assume an autogen migration is perfect as generated, alembic does not guarantee autogens that dont need manual inspection) and either identify that SQL Server doesn't have a "now()" function, or just replace server_default=text() calls with a SQL Core construct like `func.now()`.  cross-generating migrations for MSSQL + Postgresql is very doable, i did a large project with this base, but you need to always go over the migrations as there are lots of concepts that don't match across these DBs like date and time functions, unique constraints, etc.

**Reply by davetapley** [reply]
> cross-generating migrations for MSSQL + Postgresql is very doable, i did a large project with this base

That's reassuring to hear 😁 

---

I concur that checking autogenerated migrations is sensible, but I'm still not sure if/why `functions.now()` exists when `functions.current_timestamp()` is semantically the same _and_ guaranteed to be dialect agnostic 🤔

**Reply by CaselIT** [reply]
> but I'm still not sure if/why `functions.now()` exists when `functions.current_timestamp()` is semantically the same _and_ guaranteed to be dialect agnostic

I don't think there is a particular reason. This is the original issue: https://github.com/sqlalchemy/sqlalchemy/issues/943 added 14 years ago and closed by https://github.com/sqlalchemy/sqlalchemy/commit/a0ffeb546468e14fa9d99b30571f7b8f9b32f421

**Reply by zzzeek** [reply]
now() is also easier to type and current_timestamp might not have been as available then.  it would be many years before alembic would be invented and overall this is just a small inconvenience with the autogen of server defaults.

that said if "func.now()" actually rendered current_timestamp on Postgresql that would also have prevented you from having this particular issue but overall there's no guarantee of PG autogen migrations working on other DBs without manual intervention.
